flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java

```diff
@@ -0,0 +1,342 @@
+    }
+
+    /**
+     * Judge whether input RelNode meets the conditions of dimSide. If joinKeys is null means we
+     * need not consider the join keys in dim side, which already deal by dynamic partition pruning
+     * rule. If joinKeys not null means we need to judge whether joinKeys changed in dim side, if
+     * changed, this RelNode is not dim side.
+     */
+    public static boolean isDimSide(RelNode rel, @Nullable ImmutableIntList joinKeys) {
+        DppDimSideFactors dimSideFactors = new DppDimSideFactors();
+        visitDimSide(rel, dimSideFactors, joinKeys);
+        return dimSideFactors.isDimSide();
+    }
+
+    /**
+     * Visit dim side to judge whether dim side has filter condition and whether dim side's source
+     * table scan is non partitioned scan.
+     */
+    public static void visitDimSide(
+            RelNode rel, DppDimSideFactors dimSideFactors, @Nullable ImmutableIntList joinKeys) {
+        // TODO Let visitDimSide more efficient and more accurate. Like a filter on dim table or a
+        // filter for the partition field on fact table.
+        if (rel instanceof TableScan) {
+            TableScan scan = (TableScan) rel;
+            TableSourceTable table = scan.getTable().unwrap(TableSourceTable.class);
+            if (table == null) {
+                return;
+            }
+            if (!dimSideFactors.hasFilter
+                    && table.abilitySpecs() != null
+                    && table.abilitySpecs().length != 0) {
+                for (SourceAbilitySpec spec : table.abilitySpecs()) {
+                    if (spec instanceof FilterPushDownSpec) {
+                        List<RexNode> predicates = ((FilterPushDownSpec) spec).getPredicates();
+                        for (RexNode predicate : predicates) {
+                            if (isSuitableFilter(predicate)) {
+                                dimSideFactors.hasFilter = true;
+                            }
+                        }
+                    }
+                }
+            }
+            CatalogTable catalogTable = table.contextResolvedTable().getTable();
+            dimSideFactors.hasNonPartitionedScan = !catalogTable.isPartitioned();
+        } else if (rel instanceof HepRelVertex) {
+            visitDimSide(((HepRelVertex) rel).getCurrentRel(), dimSideFactors, joinKeys);
+        } else if (rel instanceof Exchange) {
+            visitDimSide(rel.getInput(0), dimSideFactors, joinKeys);
+        } else if (rel instanceof Project) {
+            // joinKeys is null means need not consider join keys.
+            if (joinKeys != null) {
+                List<RexNode> projects = ((Project) rel).getProjects();
+                ImmutableIntList inputJoinKeys = getInputIndices(projects, joinKeys);
+                if (inputJoinKeys.isEmpty()) {
+                    dimSideFactors.isSuitableJoinKey = false;
+                }
+            }
+
+            visitDimSide(rel.getInput(0), dimSideFactors, joinKeys);
```

Comment:
the joinKeys should be inputJoinKeys, consider the following situation: Project on Calc with field index chagned

Reply:
> the joinKeys should be inputJoinKeys, consider the following situation: Project on Calc with field index chagned

done!